RFE-2782: Create edge compute pool to support AWS Local Zones (#6371)
openshift/installer: docs/user/aws/install_existing_vpc_local-zones.md

            - [Create the Local Zone Subnet using AWS CLI](#create-network-subnet-res-cli)
            - [Create the Local Zone Subnet using AWS Console](#create-network-subnet-res-console)
- [Install](#install-cluster)
    - [Create the install-config.yaml](#create-config)
    - [Setting up the Edge Machine Pool](#create-config-edge-pool")
        - [Example edge pool created without customization](#create-config-edge-pool)
        - [Example edge pool with custom Instance type](#reate-config-edge-pool-example-ec2)
        - [Example edge pool with custom EBS type](#create-config-edge-pool-example-ebs)
    - [Create the cluster](#create-cluster-run)
- [Uninstall](#uninstall)
    - [Destroy the cluster](#uninstall-destroy-cluster)
    - [Destroy the Local Zone subnet](#uninstall-destroy-subnet)
    - [Destroy the VPC](#uninstall-destroy-vpc)
- [Use Cases](#use-cases)
    - [User-workload ingress traffic](#uc-exposing-ingress)
___

To install a cluster in existing VPC with Local Zone subnets, you should provision the network resources and then add the subnet IDs to the `install-config.yaml`.

> ## Pre-requisites <a name="pre-requisites"></a>

Author: yeah, I like to define the index due we don't need to keep long references, or can change the section name without needing to change the links, avoid broking it.
Thanks for your review, I will adjust it and tag you again.

Version 4.12 or later introduces a new compute pool named `edge` designed for
the remote zones. The `edge` compute pool configuration is common between
AWS Local Zone locations, but due to the limitation of resources (Instance Types
and Sizes) of the Local Zone, the default instance type created may vary
from the traditional worker pool.

The default EBS for Local Zone locations is `gp2`, different than the default worker pool.

The preferred list of instance type follows the same order of worker pools, depending
on the availability of the location, one of those instances will be chosen*:
> Note: This list can be updated over the time
- `m6i.xlarge`
- `m5.xlarge`
- `c5d.2xlarge`

The `edge`compute pool will also create new labels to help developers to
deploy their applications onto those locations. The new labels introduced are:
    - `node-role.kubernetes.io/edge=''`
    - `zone_type=local-zone`
    - `zone_group=<Local Zone Group>`

Finally, the Machine Sets created by the `edge` compute pool have `NoSchedule` taint to avoid the
regular workloads spread out on those machines, and only user workloads will be allowed to run

> when the tolerations are defined on the pod spec.

Author: For sure, we created it on the HC blog post[1]. I will paste the full example here, and adjust when updating the doc:
```
cat << EOF | oc create -f -
apiVersion: v1
kind: Namespace
metadata:
  name: lz-apps
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: lz-app-nyc-1
  namespace: lz-apps
spec:
  selector:
    matchLabels:
      app: lz-app-nyc-1
  replicas: 1
  template:
    metadata:
      labels:
        app: lz-app-nyc-1
        zone_group: ${ZONE_GROUP_NAME}
    spec:
      nodeSelector:
        zone_group: ${ZONE_GROUP_NAME}
      tolerations:
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: echoserver
          ports:
            - containerPort: 8080
---
apiVersion: v1
kind: Service 
metadata:
  name:  lz-app-nyc-1 
  namespace: lz-apps
spec:
  ports:
    - port: 80
      targetPort: 8080
      protocol: TCP
  type: NodePort
  selector:
    app: lz-app-nyc-1
EOF
```

[1] https://docs.google.com/document/d/13C-FPCgYEMsr2cyVV_pG5mpl03ZjfvXrV8d2uSJVms0/edit

> when the tolerations are defined on the pod spec.

Author: Added a new section to the `Use Case` section. Thanks for the feedback.

___

To install a cluster in existing VPC with Local Zone subnets, you should provision the network resources and then add the subnet IDs to the `install-config.yaml`.

## Pre-requisites <a name="pre-requisites"></a>

- [AWS Command Line Interface](aws-cli)
- [openshift-install >= 4.12](openshift-install)
- environment variables exported

```bash
export CLUSTER_NAME="ipi-localzones"

# AWS Region and extra Local Zone group Information
export AWS_REGION="us-west-2"
export ZONE_GROUP="us-west-2-lax-1"
export ZONE_NAME="us-west-2-lax-1a"

# VPC Information
export VPC_CIDR="10.0.0.0/16"
export VPC_SUBNETS_BITS="10"
export VPC_SUBNETS_COUNT="3"

# Local Zone Subnet information
export SUBNET_CIDR="10.0.192.0/22"
export SUBNET_NAME="${CLUSTER_NAME}-public-usw2-lax-1a"
```

## Create the Network Stack <a name="create-network"></a>

### Create the VPC <a name="create-network-vpc"></a>

The steps to install a cluster in existing VPC are [detailed on the official documentation](aws-install-vpc). You can alternatively use [the CloudFormation templates to create the Network resources](aws-install-cloudformation), which will be used in this document.

- Create the Stack

```bash

> INSTALER_URL="file:///home/mtulio/go/src/github.com/mtulio/installer"
> #INSTALER_URL="https://raw.githubusercontent.com/mtulio/installer/master"

Author: ```suggestion
INSTALER_URL="https://raw.githubusercontent.com/openshift/installer/master"
```

TPL_URL="${INSTALER_URL}/upi/aws/cloudformation/01_vpc.yaml"

aws cloudformation create-stack \
    --region ${AWS_REGION} \
    --stack-name ${CLUSTER_NAME}-vpc \
    --template-body ${TPL_URL} \
    --parameters \
        ParameterKey=VpcCidr,ParameterValue=${VPC_CIDR} \
        ParameterKey=SubnetBits,ParameterValue=${VPC_SUBNETS_BITS} \
        ParameterKey=AvailabilityZoneCount,ParameterValue=${VPC_SUBNETS_COUNT}
```

> - Check the Stack status:
> 
> ```bash
> aws cloudformation describe-stacks \
>     --region ${AWS_REGION} \
>     --stack-name ${CLUSTER_NAME}-vpc 
> ```
> 
> - Wait for the stack execution be finished: `StackStatus=CREATE_COMPLETE`.

Author: ```suggestion
- Wait for the stack execution to be created: `StackStatus=CREATE_COMPLETE`

```bash
aws cloudformation wait stack-create-complete \
    --region ${AWS_REGION} \
    --stack-name ${CLUSTER_NAME}-vpc 
```

```

  --region us-west-2 \
  --stack-name ${CLUSTER_NAME}-vpc \
  | jq -r '.Stacks[0].Outputs[] | select(.OutputKey=="PublicRouteTableId").OutputValue' )
```

- Make sure all variables have been correctly set:

```bash
echo "SUBNETS=${SUBNETS[*]}
VPC_ID=${VPC_ID}
PUBLIC_RTB_ID=${PUBLIC_RTB_ID}"
```

### Create the Local Zone subnet <a name="create-network-subnet"></a>

The following actions are required to create subnets in Local Zones:
- choose the zone group to be enabled
- opt-in the zone group

#### Opt-in Zone groups <a name="create-network-subnet-optin"></a>

##### Steps to opt-in Zone Groups using AWS CLI <a name="create-network-subnet-optin-cli"></a>

```bash
aws ec2 modify-availability-zone-group \
    --region ${AWS_REGION} \
    --group-name ${ZONE_GROUP} \
    --opt-in-status opted-in
```

##### Steps to opt-in Zone Groups using AWS Console <a name="create-network-subnet-optin-console"></a>

TBD/Is it needed?

#### Creating the Subnet <a name="create-network-subnet-res"></a>

##### Creating the subnet using AWS CloudFormation <a name="create-network-subnet-res-cfn"></a>

- Create the Stack for Local Zone subnet `us-west-2-lax-1a`

```bash

> INSTALER_URL="file:///home/mtulio/go/src/github.com/mtulio/installer"
> #INSTALER_URL="https://raw.githubusercontent.com/mtulio/installer/master"

Author: ```suggestion
INSTALER_URL="https://raw.githubusercontent.com/openshift/installer/master"
```

TPL_URL="${INSTALER_URL}/upi/aws/cloudformation/01.99_net_local-zone.yaml"

aws cloudformation create-stack \
    --region ${AWS_REGION} \
    --stack-name ${SUBNET_NAME} \
    --template-body ${TPL_URL} \
    --parameters \
        ParameterKey=VpcId,ParameterValue=${VPC_ID} \
        ParameterKey=ZoneName,ParameterValue=${ZONE_NAME} \
        ParameterKey=SubnetName,ParameterValue=${SUBNET_NAME} \
        ParameterKey=PublicSubnetCidr,ParameterValue=${SUBNET_CIDR} \
        ParameterKey=PublicRouteTableId,ParameterValue=${PUBLIC_RTB_ID}
```

> - Wait for the stack be created `StackStatus=CREATE_COMPLETE`
> 
> ```bash
> aws cloudformation describe-stacks \
>   --region ${AWS_REGION} \
>   --stack-name ${SUBNET_NAME}
> ```
> 
> - Wait for the stack execution be finished: `StackStatus=CREATE_COMPLETE`.

Author: ```suggestion
- Wait for the stack to be created `StackStatus=CREATE_COMPLETE`

```bash
aws cloudformation wait stack-create-complete \
  --region ${AWS_REGION} \
  --stack-name ${SUBNET_NAME}
```

```

- Export the Local Zone subnet ID

```bash
export SUBNET_ID=$(aws cloudformation describe-stacks \
  --region ${AWS_REGION} \
  --stack-name ${SUBNET_NAME} \
  | jq -r '.Stacks[0].Outputs[] | select(.OutputKey=="PublicSubnetIds").OutputValue' )

# Append the Local Zone Subnet ID to the Subnet List
SUBNETS+=(${SUBNET_ID})
```

- Check the total of subnets. If you choose 3 AZs to be created on the VPC stack, you should have 7 subnets on this list:

```bash
$ echo ${#SUBNETS[*]}
7
```

> ##### Creating the subnet using AWS CLI <a name="create-network-subnet-res-cli"></a>
> 
> TBD/Is it needed?

Author: Already described above
```suggestion
```

##### Creating the subnet using AWS Console <a name="create-network-subnet-res-console"></a>

TBD/Is it needed?

## Install the cluster <a name="install-cluster"></a>

> To install the cluster in existing VPC with subnets in Local Zones, you should:
> - generate the install-config.yaml, you provide yours
> - add the subnet IDs setting the option `platform.aws.subnets`
> - (option) customize the Edge Machine Pool

Author: ```suggestion
To install the cluster in existing VPC with subnets in Local Zones, you should:
- generate the `install-config.yaml`, or provide yours
- add the subnet IDs by setting the option `platform.aws.subnets`
- (optional) customize the `edge` compute pool
```

### Create the install-config.yaml <a name="create-config"></a>

Create the `install-config.yaml` providing the subnet IDs recently created:

- create the `install-config`

```bash
$ ./openshift-install create install-config --dir ${CLUSTER_NAME}
? SSH Public Key /home/user/.ssh/id_rsa.pub
? Platform aws
? Region us-west-2
? Base Domain devcluster.openshift.com
? Cluster Name ipi-localzone
? Pull Secret [? for help] **
INFO Install-Config created in: ipi-localzone     
```

- Append the subnets to the `platform.aws.subnets`:

```bash
$ echo "    subnets:"; for SB in ${SUBNETS[*]}; do echo "    - $SB"; done
    subnets:
    - subnet-0fc845d8e30fdb431
    - subnet-0a2675b7cbac2e537
    - subnet-01c0ac400e1920b47
    - subnet-0fee60966b7a93da6
    - subnet-002b48c0a91c8c641
    - subnet-093f00deb44ce81f4
    - subnet-0f85ae65796e8d107
```

### Setting up the Edge Machine Pool <a name="create-config-edge-pool"></a>

> Version 4.12 e earlier introduces a new Machine Pool designed for the edge locations.
> The Edge Machine Pool configuration is common between edge locations, but due to the limitation
> of resources (Instance Types and Sizes) of the Local Zone, the default instance type created
> may vary from the traditional worker pool.

Author: ```suggestion
Version 4.12 or later introduces a new compute pool named `edge` designed for remote zones.
The `edge` compute pool configuration is common between AWS Local Zone locations, but due to the limitation
of resources (Instance Types and Sizes) of the Local Zone, the default instance type created
may vary from the traditional worker pool.
```

The default EBS for Local Zone locations is `gp2`, different than the default worker pool.

The preferred list of instance type follows the same order of worker pools, depending
on the availability of the location, one of those instances will be chosen*:
> Note: This list can be updated over the time
- `m6i.xlarge`
- `m5.xlarge`
- `c5d.2xlarge`

> The Edge Machine Pool will also create new labels for the Machines to help developers to

Author: ```suggestion
The `edge`compute pool will also create new labels to help developers to
```

deploy their applications onto those locations. The new labels introduced are:
    - `node-role.kubernetes.io/edge=''`
    - `zone_type=local-zone`
    - `zone_group=<Local Zone Group>`

> Finally, the MachineSets created on the Edge Machine Pool have `NoSchedule` taint to avoid the
> normal workloads spread out on those machines, and only user workloads will be allowed to run
> when the tolerations are defined on the pod spec.
> 
> By default, the Edge Machine Pools will be created only when AWS Local Zone subnet IDs are added
> to the list of `platform.aws.subnets`.

Author: ```suggestion
Finally, the Machine Sets created by the `edge` compute pool have `NoSchedule` taint to avoid the
regular workloads spread out on those machines, and only user workloads will be allowed to run
when the tolerations are defined on the pod spec.

By default, the `edge` compute pool will be created only when AWS Local Zone subnet IDs are added
to the list of `platform.aws.subnets`.
```
